### ISSUE TITLE
Reset issues with the GD32VF103C

### ISSUE THREAD

**Issue by kristofmulier** [GigaDevice-Semiconductor/openocd#4]
Note: I'm working here on Windows10

I'm flashing the `GD32VF103C-START` board with OpenOCD and GDB. It works - but there are lots of warning messages, mostly related to reset issues of the chip.

<h2>1. How I flash the chip</h2>

To flash the firmware to the chip, I use a special approach such that I only need a single console. The technique is explained here:
https://embeetle.com/#embeetle-ide/manual/flash/makefile-flash-target

For this `GD32VF103C-START` board, my `.gdbinit` file looks like this:

```
# .gdbinit
# ========
# GDB commands for GD32VF103CBT6

define flash-remote
  echo \n
  echo \n--------------------------------------------------------------------------------
  echo \n  Run flash-remote function:
  echo \n  ==========================
  echo \n  flash-remote(
  echo \n      elf_file = $arg0,
  echo \n      openocd_path = $arg1,
  echo \n      probe_file = $arg2,
  echo \n      chip_file = $arg3,
  echo \n  )
  echo \n--------------------------------------------------------------------------------
  echo \n
  file $arg0
  target extended-remote | $arg1 -f $arg2 -f $arg3 -c "gdb_port pipe; log_output openocd.log"
  load
  monitor reset
  monitor resume
  monitor shutdown
  quit
end
```

Then I invoke GDB (`riscv-none-embed-gdb`) with the following flags:

```makefile
GDB_FLASHFLAGS = \
    -n \
    -batch \
    -x $(GDB_FLASHFILE) \
    -ex "flash-remote $(ELF_FILE) \"$(OCD)\" $(OPENOCD_PROBEFILE) $(OPENOCD_CHIPFILE)" \
```
With:

 - `$(GDB_FLASHFILE)` is the path to the `.gdbinit` file
 - `$(ELF_FILE)` is the path to the `.elf` file I want to flash
 - `$(OCD)` is the path to the OpenOCD executable
 - `$(OPENOCD_PROBEFILE)` is the path to the file `openocd_probe.cfg`
 - `$(OPENOCD_CHIPFILE)` is the path to the file `openocd_chip.cfg`

With the following content for `openocd_probe.cfg`:

```tcl
# openocd_probe.cfg
# =================
source [find interface/gdlink.cfg]
transport select jtag
```

And `openocd_chip.cfg`:

```tcl
# openocd_chip.cfg
# ================
adapter speed 1000
reset_config srst_only
adapter srst pulse_width 100
source [find target/gd32vf103.cfg]
```

Please note that the file `target/gd32vf103.cfg` is also slightly different (see my pull request https://github.com/GigaDevice-Semiconductor/openocd/pull/1)

<h2>2. Output</h2>

In GDB, I can see the following output:

```
$ riscv-none-embed-gdb -n -batch -x ../config/.gdbinit -ex "flash-remote application.elf openocd ../config/openocd_probe.cfg ../config/openocd_chip.cfg" 

    riscv-none-embed-gdb: warning: Couldn't determine a path for the index cache directory.
    --------------------------------------------------------------------------------
      Run flash-remote function:
      ==========================
      flash-remote(
          elf_file = application.elf,
          openocd_path = openocd,
          probe_file = ../config/openocd_probe.cfg,
          chip_file = ../config/openocd_chip.cfg,
      )
    --------------------------------------------------------------------------------
    xPack OpenOCD, x86_64 Open On-Chip Debugger 0.10.0+dev-00003-gfa63b9ee9 (2020-11-19-15:34)
    Licensed under GNU GPL v2
    For bug reports, read
    http://openocd.org/doc/doxygen/bugs.html
    jtag
    0x08000d10 in __divdf3 ()
    Loading section .init, size 0x236 lma 0x8000000
    Loading section .text, size 0x172a lma 0x8000240
    Loading section .sdata2._global_impure_ptr, size 0x4 lma 0x800196c
    Loading section .init_array, size 0x4 lma 0x8001970
    Loading section .data, size 0x68 lma 0x8001974
    Start address 0x800015c, load size 6608
    Transfer rate: 747 bytes/sec, 1321 bytes/write.
    cmsis-dap JTAG TLR_RESET
    cmsis-dap JTAG TLR_RESET
    cmsis-dap JTAG TLR_RESET
    JTAG tap: riscv.cpu tap/device found: 0x1000563d (mfg: 0x31e (Andes Technology Corporation), part: 0x0005, ver: 0x1)
    JTAG tap: auto0.tap tap/device found: 0x790007a3 (mfg: 0x3d1 (GigaDevice Semiconductor (Beijing)), part: 0x9000, ver: 0x7)
    Hart 0 didn't run coming out of reset in 2s; dmstatus=0x4f03a2; Increase the timeout with riscv set_reset_timeout_sec.
    keep_alive() was not invoked in the 1000 ms timelimit. GDB alive packet not sent! (2437 ms). Workaround: increase "set remotetimeout" in GDB

    Hart 0 unexpectedly reset!
    shutdown command invoked
    A debugging session is active.

    Inferior 1 [Remote target] will be detached.

    Quit anyway? (y or n) [answered Y; input not from terminal]
    Remote communication error.  Target disconnected.: Success.
    make: Leaving directory 'C:/Users/Gebruiker/beetle_projects/gd32vf103c_start_baremetal/build'
```

And the OpenOCD output logged in `openocd.log` is:

```
Info : Listening on port 6666 for tcl connections
Info : Listening on port 4444 for telnet connections
Info : CMSIS-DAP: SWD  Supported
Info : CMSIS-DAP: JTAG Supported
Info : CMSIS-DAP: FW Version = 2.0.0
Info : CMSIS-DAP: Interface Initialised (JTAG)
Info : SWCLK/TCK = 1 SWDIO/TMS = 1 TDI = 1 TDO = 0 nTRST = 0 nRESET = 1
Info : CMSIS-DAP: Interface ready
Info : clock speed 1000 kHz
Info : cmsis-dap JTAG TLR_RESET
Info : cmsis-dap JTAG TLR_RESET
Info : JTAG tap: riscv.cpu tap/device found: 0x1000563d (mfg: 0x31e (Andes Technology Corporation), part: 0x0005, ver: 0x1)
Info : JTAG tap: auto0.tap tap/device found: 0x790007a3 (mfg: 0x3d1 (GigaDevice Semiconductor (Beijing)), part: 0x9000, ver: 0x7)
Warn : AUTO auto0.tap - use "jtag newtap auto0 tap -irlen 5 -expected-id 0x790007a3"
Info : datacount=4 progbufsize=2
Info : Examined RISC-V core; found 1 harts
Info :  hart 0: XLEN=32, misa=0x40901105
Info : starting gdb server for riscv.cpu on pipe
Info : accepting 'gdb' connection from pipe
Info : device id = 0x19060410
Info : flash size = 128kbytes
Info : cmsis-dap JTAG TLR_RESET
Info : cmsis-dap JTAG TLR_RESET
Info : cmsis-dap JTAG TLR_RESET
Info : JTAG tap: riscv.cpu tap/device found: 0x1000563d (mfg: 0x31e (Andes Technology Corporation), part: 0x0005, ver: 0x1)
Info : JTAG tap: auto0.tap tap/device found: 0x790007a3 (mfg: 0x3d1 (GigaDevice Semiconductor (Beijing)), part: 0x9000, ver: 0x7)
Info : GD32: Flash erase ... sector erase(0 to 6)
Info : erase ok
Info : Padding image section 0 at 0x08000236 with 10 bytes
Info : Padding image section 1 at 0x0800196a with 2 bytes
Info : GD32: Flash write ... not double words to write, padding with 0xff
Info : GD32: Flash write ... half word to be prgrammed = 0x00000cef
Info : cmsis-dap JTAG TLR_RESET
Info : cmsis-dap JTAG TLR_RESET
Info : cmsis-dap JTAG TLR_RESET
Info : JTAG tap: riscv.cpu tap/device found: 0x1000563d (mfg: 0x31e (Andes Technology Corporation), part: 0x0005, ver: 0x1)
Info : JTAG tap: auto0.tap tap/device found: 0x790007a3 (mfg: 0x3d1 (GigaDevice Semiconductor (Beijing)), part: 0x9000, ver: 0x7)
Info : cmsis-dap JTAG TLR_RESET
Info : cmsis-dap JTAG TLR_RESET
Info : cmsis-dap JTAG TLR_RESET
Info : JTAG tap: riscv.cpu tap/device found: 0x1000563d (mfg: 0x31e (Andes Technology Corporation), part: 0x0005, ver: 0x1)
Info : JTAG tap: auto0.tap tap/device found: 0x790007a3 (mfg: 0x3d1 (GigaDevice Semiconductor (Beijing)), part: 0x9000, ver: 0x7)
Error: Hart 0 didn't run coming out of reset in 2s; dmstatus=0x4f03a2; Increase the timeout with riscv set_reset_timeout_sec.
Warn : keep_alive() was not invoked in the 1000 ms timelimit. GDB alive packet not sent! (2437 ms). Workaround: increase "set remotetimeout" in GDB

Info : Hart 0 unexpectedly reset!
Warn : negative acknowledgment, but no packet pending
shutdown command invoked
```

Please note the error and warning messages in the output. The flash procedure is not very stable. Sometimes it works - sometimes not.

**Comment by elfmimi** [comment]
Try this script [gd32vf103.cfg](https://gist.github.com/elfmimi/1deb9c94b0f0900ae8a9df740b62bcd6) .

You may need to replace 0x1e200a6d with 0x1000563d .

P.S. thank you for informing me about this fork. @kristofmulier

**Comment by kristofmulier** [comment]
Thank you @elfmimi for the script (more on that further down in my reply).

After the latest fix from @Xingyu Ma today (25 Nov 2020) I built OpenOCD again (`version 0.10.0 dev00002`) and it works now for the single-console-approach. The double-console-approach still fails. 

<h2>1. Single-console-approach</h2>

I've applied the single-console-approach - just like I describe in my original post above - and I can confirm it works. But it only works if I modify the `target/gd32vf103.cfg` file. As explained in my pull request (see https://github.com/GigaDevice-Semiconductor/openocd/pull/1) the `init` and `halt` commands should be removed for the single-console-approach to work.

I've also tried with the script from @elfmimi and that works too!


<h2> 2. Double-console-approach</h2>

In the double-console-approach, the GDB connection gets dropped (see issue https://github.com/GigaDevice-Semiconductor/openocd/issues/3)

This problem persists in `version 0.10.0 dev00002`, regardless of the `target/gd32vf103.cfg` script I use. I tried both the original one and the one from @elfmimi (see https://gist.github.com/elfmimi/1deb9c94b0f0900ae8a9df740b62bcd6)

**Comment by kristofmulier** [comment]
The reset issues with the `GD32VF103C-START` board seem to be fixed in `v0.10.0 dev00002` (released on Nov25). Thank you very much. There is still an issue when I try to flash with the "double-console-approach", but that issue is already described in another post: https://github.com/GigaDevice-Semiconductor/openocd/issues/3
